### ISSUE TITLE
reuse addheader destroys content

### ISSUE THREAD

**Issue by ole-tange** [fsfe/reuse-tool#338]
Before:

~~~
#!/usr/bin/env sh

# This file must be sourced in sh:
#
#   . `which env_parallel.sh`
#
# after which 'env_parallel' works
#
#
# Copyright (C) 2016-2021 Ole Tange, http://ole.tange.dk and Free
# Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>
# or write to the Free Software Foundation, Inc., 51 Franklin St,
# Fifth Floor, Boston, MA 02110-1301 USA
~~~

Running:
~~~
reuse addheader --copyright="Ole Tange, http://ole.tange.dk and Free Software and Foundation, Inc." --license="GPL-3.0-or-later" /tmp/a.sh
Successfully changed header of /tmp/a.sh
~~~

Results in:
~~~
#!/usr/bin/env sh

# Copyright (C) 2016-2021 Ole Tange, http://ole.tange.dk and Free
# SPDX-FileCopyrightText: 2021 Ole Tange, http://ole.tange.dk and Free Software and Foundation, Inc.
#
# SPDX-License-Identifier: GPL-3.0-or-later
~~~

This destroyed the usage instructions in the comment.

**Comment by mxmehl** [comment]
Yeah, we should at leats document what `addheader` does by default. Perhaps the logic has to be adjusted, and with flags that define what to do with existing headers, with copyright/license strings etc.

**Comment by nicorikken** [comment]
@ole-tange In your example the single comment block contains some documentation as well as the copyright and license information. Currently the add-header is built to replace the entire comment block, as is described in the docstring of the `find_and_replace_header` function: https://github.com/fsfe/reuse-tool/blob/9e3f4735f124c2484f4cce4c358c4d9121b520c9/src/reuse/header.py#L224-L241

To elaborate on the suggestion by @mxmehl :

We could consider the approach tools take to match the entire license text, and than use that information to replace it. But I think that would blow the complexity of the codebase out of proportion.

A more simple approach would be to consider empty lines in comment blocks as separators. In the example of this issue, the double empty line is a clear separator of content. Being able to deal with single empty lines would help deal with other exceptional cases. But as single empty lines typically exist in license statements, so then at least some text matching is required.

Another option which could work in tandem, is to identify difficult cases and not replace the entire header but only append to it. And then inform the user about it. This could be controlled using a separate commandline flag.


Thankfully we have code revision systems so most people can notice unwanted changes and can roll them back.

**Comment by ole-tange** [comment]
> Currently the add-header is built to replace the entire comment block,

You cannot call it 'addheader' and have it replace a block if you want to adhere to Principle of Least Astonishment.

'addheader' should add. Not replace and not subtract.

You should feel free to add an option called 'replaceheader' which does what 'addheader' does now. But 'addheader' should only add a header - nothing else.

**Comment by mxmehl** [comment]
We should tackle this problem soonish as the default (and mostly non-configurable) behaviour conflicts with projects' expectations when we help them to become REUSE compliant.

I currently lean towards
1. adding a flag to keep the existing header and just append/update the copyright/license information.
2. switch the default behaviour from replace -> append in the future (optionally by making the replacement an own subcommand such as `replaceheader`).

**Comment by mxmehl** [comment]
In today's maintainers call, we acknowledged that we need to find a solution to this problem and provide more choices to users of the tool. We aim to offer more options by the next release, and potentially change the default in the future with a major version bump (as it's breaking).

What we agreed on:
* Offer a new subcommand or flag to `addheader` that does actually only add REUSE information and not touch existing comment headers
* Document the different behaviours better
* No solution we find will have a perfect behaviour. The tool cannot detect e.g. ASCII boxes and perfectly fit in existing headers. For this, templates or manual scripting is needed.

What's still unclear:
* New subcommand (well, `addheader` is already taken...) or flag to `addheader`? 
* Shall we change the default in the future by making `addheader` only add info, and `replaceheader` being the current behaviour?
* How shall the tool add information? Shall it append it below the first found comment header, or prepend it (or both, with additional flags)?
* Technical detail: we should make sure that once the header has been added, the tool properly recognises its area and does not wildly add new copyright/licensing information when re-running `addheader`.
